runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -61,6 +66,42 @@ public ILogData peek(Layout layout, long globalAddress) {
         return ret == null || ret.isEmpty() ? null : ret;
     }
 
+    /**
+     * Read data from a range created from the given addresses.
+     *
+     * This will be changed when an API is added to support multiple of ranges
+     *
+     * @param layout                The layout to use for the readAll.
+     * @param globalAddresses       A set of addresses to read from.
+     * @return                      A map of addresses to committed
+     *                              addresses, hole filling if necessary.
+     */
+    @Override
+    public Map<Long, ILogData> readAll(Layout layout, Set<Long> globalAddresses) {
+        Range<Long> range = Range.encloseAll(globalAddresses);
+        long startAddress = range.lowerEndpoint();
+        long endAddress = range.upperEndpoint();
+        int numUnits = layout.getSegmentLength(startAddress);
+        log.trace("ReadAll[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
+
+        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(layout
+                .getLogUnitClient(startAddress, numUnits - 1)
+                .read(null, range)).getReadSet();
+
+        //in case of a hole, do a normal read and use its hole fill policy
```

Comment:
1. Bulk hole fill doesn't exist yet. Message only takes one address. I can make that change in a separate code change. Also, to be consistent, I will have to implement a fillAll() for the hole fill policy classes.
2. I don't like using TreeMaps instead of Map in the method signatures. In my implementation, I can certainly make sure that I use a TreeMap to hold the content.